src/cpu_shader/euler.rs

```diff
@@ -0,0 +1,277 @@
+// Copyright 2023 the Vello Authors
+// SPDX-License-Identifier: Apache-2.0 OR MIT OR Unlicense
+
+//! Utility functions for Euler Spiral based stroke expansion.
+
+// Use the same constants as the f64 version.
+#![allow(clippy::excessive_precision)]
+
+use super::util::Vec2;
+use std::f32::consts::FRAC_PI_4;
+
+#[derive(Debug)]
+pub struct CubicParams {
+    pub th0: f32,
+    pub th1: f32,
+    pub d0: f32,
+    pub d1: f32,
+}
+
+#[derive(Debug)]
+pub struct EulerParams {
+    pub th0: f32,
+    pub th1: f32,
+    pub k0: f32,
+    pub k1: f32,
+    pub ch: f32,
+}
+
+#[derive(Debug)]
+pub struct EulerSeg {
+    pub p0: Vec2,
+    pub p1: Vec2,
+    pub params: EulerParams,
+}
+
+impl CubicParams {
+    /// Compute parameters from endpoints and derivatives.
+    pub fn from_points_derivs(p0: Vec2, p1: Vec2, q0: Vec2, q1: Vec2, dt: f32) -> Self {
+        let chord = p1 - p0;
+        // Robustness note: we must protect this function from being called when the
+        // chord length is (near-)zero.
+        let scale = dt / chord.length_squared();
+        let h0 = Vec2::new(
+            q0.x * chord.x + q0.y * chord.y,
+            q0.y * chord.x - q0.x * chord.y,
+        );
+        let th0 = h0.atan2();
+        let d0 = h0.length() * scale;
+        let h1 = Vec2::new(
+            q1.x * chord.x + q1.y * chord.y,
+            q1.x * chord.y - q1.y * chord.x,
+        );
+        let th1 = h1.atan2();
+        let d1 = h1.length() * scale;
+        // Robustness note: we may want to clamp the magnitude of the angles to
+        // a bit less than pi. Perhaps here, perhaps downstream.
+        CubicParams { th0, th1, d0, d1 }
+    }
+
+    // Estimated error of GH to Euler spiral
+    //
+    // Return value is normalized to chord - to get actual error, multiply
+    // by chord.
+    pub fn est_euler_err(&self) -> f32 {
+        // Potential optimization: work with unit vector rather than angle
+        let cth0 = self.th0.cos();
+        let cth1 = self.th1.cos();
+        if cth0 * cth1 < 0.0 {
+            // Rationale: this happens when fitting a cusp or near-cusp with
+            // a near 180 degree u-turn. The actual ES is bounded in that case.
+            // Further subdivision won't reduce the angles if actually a cusp.
+            return 2.0;
```

Comment:
As we discussed briefly, I would elaborate on the choice of `2.0` here: what is the expected downstream effect and why is that acceptable.

Reply:
Maybe add a short sentence about the effect of `2.0` vs another number like `1.0`, e.g. "Returning an error value of 2.0 terminates the subdivisions because so and so".